File: SegmentTree/RangeFrequencyQueries/range.py
class RangeFreqQuery:

    def __init__(self, arr):
        self.n = len(arr)
        self.freqTree = [{} for _ in range(4 * self.n)]

        self.buildFreqTree(
            0, arr, 0, self.n - 1
        )

    def addUnorderedMap(self, mp1, mp2):
        mp = {}

        for ele, freq in mp1.items():
            mp[ele] = freq

            if ele in mp2:
                mp[ele] += mp2[ele]

        for ele, freq in mp2.items():
            if ele not in mp:
                mp[ele] = freq

        return mp

    def buildFreqTree(self, i, arr, lo, hi):

        if lo == hi:
            self.freqTree[i][arr[lo]] = 1
            return

        mid = lo + (hi - lo) // 2

        self.buildFreqTree(
            2*i + 1, arr, lo, mid
        )

        self.buildFreqTree(
            2*i + 2, arr, mid + 1, hi
        )

        self.freqTree[i] = self.addUnorderedMap(
            self.freqTree[2*i + 1],
            self.freqTree[2*i + 2]
        )

    def doQuery(self, l, r, i, lo, hi, val):

        if l > hi or r < lo:
            return 0

        if l <= lo and hi <= r:
            return self.freqTree[i].get(val, 0)

        mid = lo + (hi - lo) // 2

        left = self.doQuery(
            l, r, 2*i + 1, lo, mid, val
        )

        right = self.doQuery(
            l, r, 2*i + 2, mid + 1, hi, val
        )

        return left + right

    def query(self, left, right, value):
        return self.doQuery(
            left, right, 0, 0, self.n - 1, value
        )

File: SegmentTree/RangeFrequencyQueries/range.cpp
#include <bits/stdc++.h>
using namespace std;

class RangeFreqQuery {
public:
    typedef unordered_map<int,int> type;
    vector<type>freqTree;
    int n;
    type addUnorderedMap(type mp1, type mp2){
        type mp;
        for(auto Pair: mp1){
            int ele = Pair.first;
            mp[ele] = mp1[ele];
            if(mp2.find(ele)!=mp2.end()){
                mp[ele] += mp2[ele];
                mp2.erase(ele);
            }
        }
        for(auto Pair: mp2){
            int ele = Pair.first;
            mp[ele] = mp2[ele];
        }
        return mp;
    }
    void buildFreqTree(int i, vector<type>& freqTree, vector<int>& arr, int lo, int hi){
        if(lo==hi){
            unordered_map<int,int>mp;
            mp[arr[lo]] = 1;
            freqTree[i] = mp;
            return;
        }
        int mid = lo+(hi-lo)/2;
        buildFreqTree(2*i+1,freqTree,arr,lo,mid);
        buildFreqTree(2*i+2,freqTree,arr,mid+1,hi);
        freqTree[i] = addUnorderedMap(freqTree[2*i+1],freqTree[2*i+2]);
    }
    RangeFreqQuery(vector<int>& arr) {
        this->n = arr.size();
        freqTree.resize(4*n);
        buildFreqTree(0,freqTree,arr,0,n-1);
    }
    int doQuery(int l, int r, int i, int lo, int hi, int val){
        if(l>hi || r<lo)return 0;
        else if(l<=lo && hi<=r){
            if(freqTree[i].find(val)==freqTree[i].end())return 0;
            return freqTree[i][val];
        }
        int mid = lo+(hi-lo)/2;
        int left = doQuery(l,r,2*i+1,lo,mid,val);
        int right =doQuery(l,r,2*i+2,mid+1,hi,val);
        return left+right;
    }
    int query(int left, int right, int value) {
        return doQuery(left,right,0,0,n-1,value);
    }
};

/**
 * Your RangeFreqQuery object will be instantiated and called as such:
 * RangeFreqQuery* obj = new RangeFreqQuery(arr);
 * int param_1 = obj->query(left,right,value);
 */